mons bind to default port after upgrade from legacy, causing upgrade failure

<!-- **Are you in the right place?**
1. For issues or feature requests, please create an issue in this repository.
2. For general technical and non-technical questions, we are happy to help you on our [Rook.io Slack](https://Rook-io.slack.com).
3. Did you already search the existing open issues for anything similar? -->

**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
In 0.9 and previous we were using port 6790 as a non-default port for the mons.  In master we are now using the default port of 6789. The upgrade from 0.9 to master is failing due to the following sequence:
- 0.9 starts mons on port 6790
- The cluster is upgraded to master
- The new mon pod spec correctly configures port 6790 for the upgraded mon
- The mon is actually listening on port 6789 inside the pod because there is no ceph.conf configuration that indicates it should listen on a non-default port.
- The mons cannot form quorum and the operator times out connecting to the mons after upgrade.

**Expected behavior:**
Mons should be listening on port 6790 after the upgrade.

@BlaineEXE can we generate the config for mons to listen on the non-default port? In 0.9 the config generated the following, but we can hopefully do this with a command line option now.
```
[mon.a]
public bind addr = 172.17.0.6:6790
```

**How to reproduce it (minimal and precise):**
<!-- Please let us know any circumstances for reproduction of your bug. -->

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)):

@travisn what is the process for addressing this:
```
=== checking if vendor deps changed

# Gopkg.lock is out of sync:

k8s.io/apimachinery/pkg/api/equality: imported or required, but missing from Gopkg.lock's input-imports

make: *** [go.vendor.check] Error 1

build/makelib/golang.mk:181: recipe for target 'go.vendor.check' failed

script returned exit code 1
```

--- a/pkg/daemon/ceph/client/auth.go
+++ b/pkg/daemon/ceph/client/auth.go
@@ -59,7 +59,6 @@ func AuthGetKey(context *clusterd.Context, clusterName, name string) (string, er
 
 // AuthGetOrCreateKey gets or creates the key for the given user.
 func AuthGetOrCreateKey(context *clusterd.Context, clusterName, name string, caps []string) (string, error) {
-
 	args := append([]string{"auth", "get-or-create-key", name}, caps...)
 	buf, err := ExecuteCephCommand(context, clusterName, args)
 	if err != nil {
@@ -67,6 +66,16 @@ func AuthGetOrCreateKey(context *clusterd.Context, clusterName, name string, cap
 	}
 
 	return parseAuthKey(buf)
+}
+
+// AuthUpdateCaps updates the capabilities for the given user.
+func AuthUpdateCaps(context *clusterd.Context, clusterName, name string, caps []string) error {
+	args := append([]string{"auth", "caps", name}, caps...)
+	_, err := ExecuteCephCommand(context, clusterName, args)
+	if err != nil {
+		return fmt.Errorf("failed to update caps for %s. %+v", name, err)
+	}
+	return err
 }
 
 // AuthDelete will delete the given user.

--- a/pkg/operator/ceph/cluster/mon/health.go
+++ b/pkg/operator/ceph/cluster/mon/health.go
@@ -324,7 +324,7 @@ func (c *Cluster) removeMon(daemonName string) error {
 		delete(c.mapping.Node, daemonName)
 		// if node->port "mapping" has been created, decrease or delete it
 		if port, ok := c.mapping.Port[nodeName]; ok {
-			if port == DefaultPort {
+			if port == DefaultMsgr1Port {
 				delete(c.mapping.Port, nodeName)
 			}
 			// don't clean up if a node port is higher than the default port, other

--- a/pkg/operator/ceph/cluster/mon/health_test.go
+++ b/pkg/operator/ceph/cluster/mon/health_test.go
@@ -38,6 +38,7 @@ import (
 )
 
 func TestCheckHealth(t *testing.T) {
+
 	var deploymentsUpdated *[]*apps.Deployment
 	updateDeploymentAndWait, deploymentsUpdated = testopk8s.UpdateDeploymentAndWaitStub()
 
@@ -64,7 +65,7 @@ func TestCheckHealth(t *testing.T) {
 		Name:    "node0",
 		Address: "",
 	}
-	c.mapping.Port["node0"] = DefaultPort
+	c.mapping.Port["node0"] = DefaultMsgr1Port
 	c.maxMonID = 4
 
 	err := c.checkHealth()
@@ -117,7 +118,7 @@ func TestCheckHealthNotFound(t *testing.T) {
 	c.mapping.Node["b"] = &NodeInfo{
 		Name: "node0",
 	}
-	c.mapping.Port["node0"] = DefaultPort
+	c.mapping.Port["node0"] = DefaultMsgr1Port
 	c.maxMonID = 4
 
 	c.saveMonConfig()

--- a/pkg/operator/ceph/cluster/mon/mon.go
+++ b/pkg/operator/ceph/cluster/mon/mon.go
@@ -66,16 +66,20 @@ const (
 	DefaultMonCount = 3
 	// MaxMonCount Maximum allowed mon count for a cluster
 	MaxMonCount = 9
-	// Msgr2port is the listening port of the messenger v2 protocol
-	Msgr2port int32 = 3300
+
+	// DefaultMsgr1Port is the default port Ceph mons use to communicate amongst themselves prior
+	// to Ceph Nautilus.
+	DefaultMsgr1Port int32 = 6789
+	// DefaultMsgr2Port is the listening port of the messenger v2 protocol introduced in Ceph
+	// Nautilus. In Nautilus and a few Ceph releases after, Ceph can use both v1 and v2 protocols.
+	DefaultMsgr2Port int32 = 3300
+
 	// minimum amount of memory in MB to run the pod
 	cephMonPodMinimumMemory uint64 = 1024
 )
 
 var (
 	logger = capnslog.NewPackageLogger("github.com/rook/rook", "op-mon")
-	// DefaultPort is the default port Ceph mons use to communicate amongst themselves.
-	DefaultPort int32 = 6789
 )
 
 // Cluster represents the Rook and environment configuration settings needed to set up Ceph mons.
@@ -317,7 +321,7 @@ func (c *Cluster) newMonConfig(monID int) *monConfig {
 	return &monConfig{
 		ResourceName: resourceName(daemonName),
 		DaemonName:   daemonName,
-		Port:         DefaultPort,
+		Port:         DefaultMsgr1Port,
 		DataPathMap: config.NewStatefulDaemonDataPathMap(
 			c.dataDirHostPath, dataDirRelativeHostPath(daemonName), config.MonType, daemonName),
 	}

--- a/pkg/operator/ceph/cluster/mon/mon_test.go
+++ b/pkg/operator/ceph/cluster/mon/mon_test.go
@@ -62,7 +62,7 @@ func testGenMonConfig(monID string) *monConfig {
 	return &monConfig{
 		ResourceName: "rook-ceph-" + moniker, // rook-ceph-mon-A or rook-ceph-mon#
 		DaemonName:   monID,                  // A or mon#
-		Port:         DefaultPort,
+		Port:         DefaultMsgr1Port,
 		PublicIP:     fmt.Sprintf("2.4.6.%d", index+1),
 		// dataDirHostPath assumed to be /var/lib/rook
 		DataPathMap: config.NewStatefulDaemonDataPathMap(
@@ -142,6 +142,7 @@ func TestResourceName(t *testing.T) {
 }
 
 func TestStartMonPods(t *testing.T) {
+
 	namespace := "ns"
 	context := newTestStartCluster(namespace)
 	c := newCluster(context, namespace, false, true, v1.ResourceRequirements{})
@@ -160,6 +161,7 @@ func TestStartMonPods(t *testing.T) {
 }
 
 func TestOperatorRestart(t *testing.T) {
+
 	namespace := "ns"
 	context := newTestStartCluster(namespace)
 	c := newCluster(context, namespace, false, true, v1.ResourceRequirements{})
@@ -184,6 +186,7 @@ func TestOperatorRestart(t *testing.T) {
 
 // safety check that if hostNetwork is used no changes occur on an operator restart
 func TestOperatorRestartHostNetwork(t *testing.T) {
+
 	namespace := "ns"
 	context := newTestStartCluster(namespace)
 

--- a/pkg/operator/ceph/cluster/mon/service.go
+++ b/pkg/operator/ceph/cluster/mon/service.go
@@ -22,14 +22,13 @@ import (
 
 	"github.com/rook/rook/pkg/operator/k8sutil"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
 func (c *Cluster) createService(mon *monConfig) (string, error) {
 	labels := c.getLabels(mon.DaemonName)
-	s := &v1.Service{
+	svcDef := &v1.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:   mon.ResourceName,
 			Labels: labels,
@@ -37,38 +36,34 @@ func (c *Cluster) createService(mon *monConfig) (string, error) {
 		Spec: v1.ServiceSpec{
 			Ports: []v1.ServicePort{
 				{
-					Name:       mon.ResourceName,
-					Port:       mon.Port,
-					TargetPort: intstr.FromInt(int(mon.Port)),
+					Name: "msgr1",
+					Port: mon.Port,
+					// --public-bind-addr=IP with no IP:port has the mon listen on port 6789
+					// regardless of what port the mon advertises (--public-addr) to the outside.
+					TargetPort: intstr.FromInt(int(DefaultMsgr1Port)),
 					Protocol:   v1.ProtocolTCP,
 				},
 			},
 			Selector: labels,
 		},
 	}
-	k8sutil.SetOwnerRef(c.context.Clientset, c.Namespace, &s.ObjectMeta, &c.ownerRef)
+	k8sutil.SetOwnerRef(c.context.Clientset, c.Namespace, &svcDef.ObjectMeta, &c.ownerRef)
 	if c.HostNetwork {
-		s.Spec.ClusterIP = v1.ClusterIPNone
+		svcDef.Spec.ClusterIP = v1.ClusterIPNone
 	}
 
 	// If deploying Nautilus or newer we need a new port for the monitor service
 	if c.clusterInfo.CephVersion.IsAtLeastNautilus() {
-		addServicePort(s, "msgr2", Msgr2port)
+		addServicePort(svcDef, "msgr2", DefaultMsgr2Port)
 	}
 
-	s, err := c.context.Clientset.CoreV1().Services(c.Namespace).Create(s)
+	s, err := k8sutil.CreateOrUpdateService(c.context.Clientset, c.Namespace, svcDef)
 	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return "", fmt.Errorf("failed to create mon service. %+v", err)
-		}
-		s, err = c.context.Clientset.CoreV1().Services(c.Namespace).Get(mon.ResourceName, metav1.GetOptions{})
-		if err != nil {
-			return "", fmt.Errorf("failed to get mon %s service ip. %+v", mon.ResourceName, err)
-		}
+		return "", fmt.Errorf("failed to create service for mon %s. %+v", mon.DaemonName, err)
 	}
 
 	if s == nil {
-		logger.Warningf("service ip not found for mon %s. this better be a test", mon.ResourceName)
+		logger.Errorf("service ip not found for mon %s. if this is not a unit test, this is an error", mon.ResourceName)
 		return "", nil
 	}
 
@@ -76,7 +71,7 @@ func (c *Cluster) createService(mon *monConfig) (string, error) {
 	// however it's interesting to show that monitors can be addressed via 2 different ports
 	// in the end the service has msgr1 and msgr2 ports configured so it's not entirely wrong
 	if c.clusterInfo.CephVersion.IsAtLeastNautilus() {
-		logger.Infof("mon %s endpoint are [v2:%s:%s,v1:%s:%d]", mon.DaemonName, s.Spec.ClusterIP, strconv.Itoa(int(Msgr2port)), s.Spec.ClusterIP, mon.Port)
+		logger.Infof("mon %s endpoint are [v2:%s:%s,v1:%s:%d]", mon.DaemonName, s.Spec.ClusterIP, strconv.Itoa(int(DefaultMsgr2Port)), s.Spec.ClusterIP, mon.Port)
 	} else {
 		logger.Infof("mon %s endpoint is %s:%d", mon.DaemonName, s.Spec.ClusterIP, mon.Port)
 	}

--- a/pkg/operator/ceph/cluster/mon/spec.go
+++ b/pkg/operator/ceph/cluster/mon/spec.go
@@ -164,7 +164,11 @@ func (c *Cluster) makeMonDaemonContainer(monConfig *monConfig) v1.Container {
 		Args: append(
 			opspec.DaemonFlags(c.clusterInfo, monConfig.DaemonName),
 			"--foreground",
+			// If the mon is already in the monmap, when the port is left off of --public-addr,
+			// it will still advertise on the previous port b/c monmap is saved to mon database.
 			config.NewFlag("public-addr", monConfig.PublicIP),
+			// Opposite of the above, --public-bind-addr will *not* still advertise on the previous
+			// port, which makes sense because this is the pod IP, which changes with every new pod.
 			config.NewFlag("public-bind-addr", opspec.ContainerEnvVarReference(podIPEnvVar)),
 		),
 		Image:           c.spec.CephVersion.Image,

--- a/pkg/operator/ceph/config/keyring/store.go
+++ b/pkg/operator/ceph/config/keyring/store.go
@@ -69,7 +69,16 @@ func (k *SecretStore) GenerateKey(resourceName, user string, access []string) (s
 	// get-or-create-key for the user account
 	key, err := client.AuthGetOrCreateKey(k.context, k.namespace, user, access)
 	if err != nil {
-		return "", fmt.Errorf("failed to get or create auth key for %s. %+v", user, err)
+		logger.Infof("Error getting or creating key for %s. "+
+			"Attempting to update capabilities in case the user already exists. %+v", user, err)
+		uErr := client.AuthUpdateCaps(k.context, k.namespace, user, access)
+		if uErr != nil {
+			return "", fmt.Errorf("failed to get, create, or update auth key for %s. %+v", user, err)
+		}
+		key, uErr = client.AuthGetKey(k.context, k.namespace, user)
+		if uErr != nil {
+			return "", fmt.Errorf("failed to get key after updating existing auth capabilities for %s. %+v", user, err)
+		}
 	}
 	return key, nil
 }

--- a/pkg/operator/ceph/object/rgw.go
+++ b/pkg/operator/ceph/object/rgw.go
@@ -62,7 +62,7 @@ func (c *clusterConfig) createOrUpdate(update bool) error {
 	if err == nil && exists {
 		if !update {
 			logger.Infof("object store %s exists in namespace %s", c.store.Name, c.store.Namespace)
-			return c.startRGWPods(false)
+			return c.startRGWPods(update)
 		}
 		logger.Infof("object store %s exists in namespace %s. checking for updates", c.store.Name, c.store.Namespace)
 	}

--- a/pkg/operator/ceph/object/spec.go
+++ b/pkg/operator/ceph/object/spec.go
@@ -50,20 +50,30 @@ func (c *clusterConfig) startDeployment() (*apps.Deployment, error) {
 	k8sutil.SetOwnerRefs(c.context.Clientset, c.store.Namespace, &d.ObjectMeta, c.ownerRefs)
 
 	logger.Debugf("starting rgw deployment: %+v", d)
-	deployment, err := c.context.Clientset.Apps().Deployments(c.store.Namespace).Create(d)
-	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return nil, fmt.Errorf("failed to create rgw deployment %s: %+v", c.instanceName(), err)
-		}
-		logger.Infof("deployment for rgw %s already exists. updating if needed", c.instanceName())
-		// There may be a *lot* of rgws, and they are stateless, so don't bother waiting until the
-		// entire deployment is updated to move on.
-		deployment, err = c.context.Clientset.Apps().Deployments(c.store.Namespace).Update(d)
-		if err != nil {
-			return nil, fmt.Errorf("failed to update rgw deployment %s. %+v", c.instanceName(), err)
+	deployment, err := c.context.Clientset.Apps().Deployments(c.store.Namespace).Get(d.Name, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to see if rgw deployment %s already exists. %+v", d.Name, err)
+	} else if err == nil {
+		// deployment exists
+		var uErr error
+		deployment, uErr = c.context.Clientset.Apps().Deployments(c.store.Namespace).Update(d)
+		if uErr != nil {
+			// may fail to update when labels have changed on the deployment and thus the label selector
+			// in this case we can try to delete the deployment and recreate
+			dErr := c.context.Clientset.Apps().Deployments(c.store.Namespace).Delete(d.Name, &metav1.DeleteOptions{})
+			if dErr != nil {
+				return nil, fmt.Errorf("failed to delete existing rgw deployment %s as part of update attempt. %+v", d.Name, dErr)
+			}
+		} else {
+			return deployment, uErr
 		}
 	}
-	return deployment, nil
+	// err != nil && isNotFound  or  err == nil && update failed, causing earlier dep to be deleted
+	deployment, err = c.context.Clientset.Apps().Deployments(c.store.Namespace).Create(d)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create rgw deployment %s: %+v", c.instanceName(), err)
+	}
+	return deployment, err
 }
 
 func (c *clusterConfig) startDaemonset() (*apps.DaemonSet, error) {

--- a/pkg/operator/k8sutil/daemonset.go
+++ b/pkg/operator/k8sutil/daemonset.go
@@ -1,0 +1,52 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package k8sutil
+
+import (
+	"fmt"
+
+	apps "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// CreateDaemonSet creates
+func CreateDaemonSet(name, namespace string, clientset kubernetes.Interface, ds *apps.DaemonSet) error {
+	_, err := clientset.Apps().DaemonSets(namespace).Create(ds)
+	if err != nil {
+		if k8serrors.IsAlreadyExists(err) {
+			_, err = clientset.Apps().DaemonSets(namespace).Update(ds)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to start %s daemonset: %v\n%v", name, err, ds)
+		}
+	}
+	return err
+}
+
+// DeleteDaemonset makes a best effort at deleting a daemonset and its pods, then waits for them to be deleted
+func DeleteDaemonset(clientset kubernetes.Interface, namespace, name string) error {
+	deleteAction := func(options *metav1.DeleteOptions) error {
+		return clientset.Apps().DaemonSets(namespace).Delete(name, options)
+	}
+	getAction := func() error {
+		_, err := clientset.Apps().DaemonSets(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	return deleteResourceAndWait(namespace, name, "daemonset", deleteAction, getAction)
+}

--- a/pkg/operator/k8sutil/deployment.go
+++ b/pkg/operator/k8sutil/deployment.go
@@ -68,7 +68,7 @@ func WaitForDeploymentImage(clientset kubernetes.Interface, namespace, label, co
 
 		if matches == len(deployments.Items) && matches > 0 {
 			logger.Infof("all %d %s deployments are on image %s", matches, label, desiredImage)
-			break
+			return nil
 		}
 
 		if len(deployments.Items) == 0 {
@@ -78,7 +78,7 @@ func WaitForDeploymentImage(clientset kubernetes.Interface, namespace, label, co
 		}
 		time.Sleep(time.Duration(sleepTime) * time.Second)
 	}
-	return nil
+	return fmt.Errorf("failed to wait for image %s in label %s", desiredImage, label)
 }
 
 // UpdateDeploymentAndWait updates a deployment and waits until it is running to return. It will

--- a/pkg/operator/k8sutil/deployment.go
+++ b/pkg/operator/k8sutil/deployment.go
@@ -13,6 +13,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+
 package k8sutil
 
 import (
@@ -129,4 +130,17 @@ func GetDeployments(clientset kubernetes.Interface, namespace, labelSelector str
 		return nil, fmt.Errorf("failed to list deployments with labelSelector %s: %v", labelSelector, err)
 	}
 	return deployments, nil
+}
+
+// DeleteDeployment makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
+func DeleteDeployment(clientset kubernetes.Interface, namespace, name string) error {
+	logger.Debugf("removing %s deployment if it exists", name)
+	deleteAction := func(options *metav1.DeleteOptions) error {
+		return clientset.Apps().Deployments(namespace).Delete(name, options)
+	}
+	getAction := func() error {
+		_, err := clientset.Apps().Deployments(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	return deleteResourceAndWait(namespace, name, "deployment", deleteAction, getAction)
 }

--- a/pkg/operator/k8sutil/deployment.go
+++ b/pkg/operator/k8sutil/deployment.go
@@ -45,43 +45,6 @@ func GetDeploymentSpecImage(clientset kubernetes.Interface, d apps.Deployment, c
 	return image, nil
 }
 
-func WaitForDeploymentImage(clientset kubernetes.Interface, namespace, label, container string, initContainer bool, desiredImage string) error {
-
-	sleepTime := 3
-	attempts := 30
-	for i := 0; i < attempts; i++ {
-		deployments, err := clientset.Apps().Deployments(namespace).List(metav1.ListOptions{LabelSelector: label})
-		if err != nil {
-			return fmt.Errorf("failed to list deployments with label %s. %v", label, err)
-		}
-
-		matches := 0
-		for _, d := range deployments.Items {
-			image, err := GetDeploymentSpecImage(clientset, d, container, initContainer)
-			if err != nil {
-				logger.Infof("failed to get image for deployment %s. %+v", d.Name, err)
-				continue
-			}
-			if image == desiredImage {
-				matches++
-			}
-		}
-
-		if matches == len(deployments.Items) && matches > 0 {
-			logger.Infof("all %d %s deployments are on image %s", matches, label, desiredImage)
-			return nil
-		}
-
-		if len(deployments.Items) == 0 {
-			logger.Infof("waiting for at least one deployment to start to see the version")
-		} else {
-			logger.Infof("%d/%d %s deployments match image %s", matches, len(deployments.Items), label, desiredImage)
-		}
-		time.Sleep(time.Duration(sleepTime) * time.Second)
-	}
-	return fmt.Errorf("failed to wait for image %s in label %s", desiredImage, label)
-}
-
 // UpdateDeploymentAndWait updates a deployment and waits until it is running to return. It will
 // error if the deployment does not exist to be updated or if it takes too long.
 func UpdateDeploymentAndWait(context *clusterd.Context, deployment *apps.Deployment, namespace string) (*v1.Deployment, error) {

--- a/pkg/operator/k8sutil/k8sutil.go
+++ b/pkg/operator/k8sutil/k8sutil.go
@@ -21,8 +21,11 @@ import (
 	"crypto/md5"
 	"fmt"
 	"strings"
+	"time"
 
 	"github.com/coreos/pkg/capnslog"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/kubernetes/pkg/util/version"
@@ -84,4 +87,44 @@ func TruncateNodeName(format, nodeName string) string {
 		nodeName = hashed
 	}
 	return fmt.Sprintf(format, nodeName)
+}
+
+// deleteResourceAndWait will delete a resource, then wait for it to be purged from the system
+func deleteResourceAndWait(namespace, name, resourceType string,
+	deleteAction func(*metav1.DeleteOptions) error,
+	getAction func() error,
+) error {
+	var gracePeriod int64
+	propagation := metav1.DeletePropagationForeground
+	options := &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod, PropagationPolicy: &propagation}
+
+	// Delete the resource if it exists
+	logger.Infof("removing %s %s if it exists", resourceType, name)
+	err := deleteAction(options)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to delete %s. %+v", name, err)
+		}
+		return nil
+	}
+	logger.Infof("Removed %s %s", resourceType, name)
+
+	// wait for the resource to be deleted
+	sleepTime := 2 * time.Second
+	for i := 0; i < 30; i++ {
+		// check for the existence of the resource
+		err = getAction()
+		if err != nil {
+			if errors.IsNotFound(err) {
+				logger.Infof("confirmed %s does not exist", name)
+				return nil
+			}
+			return fmt.Errorf("failed to get %s. %+v", name, err)
+		}
+
+		logger.Infof("%s still found. waiting...", name)
+		time.Sleep(sleepTime)
+	}
+
+	return fmt.Errorf("gave up waiting for %s pods to be terminated", name)
 }

--- a/pkg/operator/k8sutil/name.go
+++ b/pkg/operator/k8sutil/name.go
@@ -10,6 +10,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+
 package k8sutil
 
 import "fmt"

--- a/pkg/operator/k8sutil/pod.go
+++ b/pkg/operator/k8sutil/pod.go
@@ -23,13 +23,11 @@ import (
 	"os"
 	"path"
 	"strings"
-	"time"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/resource"
 )
 
@@ -220,83 +218,6 @@ func GetPodLog(clientset kubernetes.Interface, namespace string, labelSelector s
 	return "", fmt.Errorf("did not find any pods with label %s", labelSelector)
 }
 
-// DeleteDeployment makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
-func DeleteDeployment(clientset kubernetes.Interface, namespace, name string) error {
-	logger.Debugf("removing %s deployment if it exists", name)
-	deleteAction := func(options *metav1.DeleteOptions) error {
-		return clientset.Apps().Deployments(namespace).Delete(name, options)
-	}
-	getAction := func() error {
-		_, err := clientset.Apps().Deployments(namespace).Get(name, metav1.GetOptions{})
-		return err
-	}
-	return deleteResourceAndWait(namespace, name, "deployment", deleteAction, getAction)
-}
-
-// DeleteReplicaSet makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
-func DeleteReplicaSet(clientset kubernetes.Interface, namespace, name string) error {
-	deleteAction := func(options *metav1.DeleteOptions) error {
-		return clientset.Apps().ReplicaSets(namespace).Delete(name, options)
-	}
-	getAction := func() error {
-		_, err := clientset.Apps().ReplicaSets(namespace).Get(name, metav1.GetOptions{})
-		return err
-	}
-	return deleteResourceAndWait(namespace, name, "replicaset", deleteAction, getAction)
-}
-
-// DeleteDaemonset makes a best effort at deleting a daemonset and its pods, then waits for them to be deleted
-func DeleteDaemonset(clientset kubernetes.Interface, namespace, name string) error {
-	logger.Infof("removing %s daemonset if it exists", name)
-	deleteAction := func(options *metav1.DeleteOptions) error {
-		return clientset.Apps().DaemonSets(namespace).Delete(name, options)
-	}
-	getAction := func() error {
-		_, err := clientset.Apps().DaemonSets(namespace).Get(name, metav1.GetOptions{})
-		return err
-	}
-	return deleteResourceAndWait(namespace, name, "daemonset", deleteAction, getAction)
-}
-
-// deleteResourceAndWait will delete a resource, then wait for it to be purged from the system
-func deleteResourceAndWait(namespace, name, resourceType string,
-	deleteAction func(*metav1.DeleteOptions) error,
-	getAction func() error) error {
-
-	var gracePeriod int64
-	propagation := metav1.DeletePropagationForeground
-	options := &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod, PropagationPolicy: &propagation}
-
-	// Delete the resource if it exists
-	err := deleteAction(options)
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("failed to delete %s. %+v", name, err)
-		}
-		return nil
-	}
-	logger.Infof("Removed %s %s", resourceType, name)
-
-	// wait for the resource to be deleted
-	sleepTime := 2 * time.Second
-	for i := 0; i < 30; i++ {
-		// check for the existence of the resource
-		err = getAction()
-		if err != nil {
-			if errors.IsNotFound(err) {
-				logger.Infof("confirmed %s does not exist", name)
-				return nil
-			}
-			return fmt.Errorf("failed to get %s. %+v", name, err)
-		}
-
-		logger.Infof("%s still found. waiting...", name)
-		time.Sleep(sleepTime)
-	}
-
-	return fmt.Errorf("gave up waiting for %s pods to be terminated", name)
-}
-
 // ClusterDaemonEnvVars Environment variables used by storage cluster daemon
 func ClusterDaemonEnvVars(image string) []v1.EnvVar {
 	return []v1.EnvVar{
@@ -321,9 +242,4 @@ func ClusterDaemonEnvVars(image string) []v1.EnvVar {
 		// Kubernetes will set this variable to 0 or equal to limits.cpu if set
 		{Name: "POD_CPU_REQUEST", ValueFrom: &v1.EnvVarSource{ResourceFieldRef: &v1.ResourceFieldSelector{Resource: "requests.cpu"}}},
 	}
-}
-
-// CephVersionName is the name of the ceph version
-func CephVersionName(version string) v1.EnvVar {
-	return v1.EnvVar{Name: "ROOK_CEPH_VERSION_NAME", Value: version}
 }

--- a/pkg/operator/k8sutil/replicaset.go
+++ b/pkg/operator/k8sutil/replicaset.go
@@ -1,0 +1,34 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package k8sutil
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// DeleteReplicaSet makes a best effort at deleting a deployment and its pods, then waits for them to be deleted
+func DeleteReplicaSet(clientset kubernetes.Interface, namespace, name string) error {
+	deleteAction := func(options *metav1.DeleteOptions) error {
+		return clientset.Apps().ReplicaSets(namespace).Delete(name, options)
+	}
+	getAction := func() error {
+		_, err := clientset.Apps().ReplicaSets(namespace).Get(name, metav1.GetOptions{})
+		return err
+	}
+	return deleteResourceAndWait(namespace, name, "replicaset", deleteAction, getAction)
+}

--- a/pkg/operator/k8sutil/service.go
+++ b/pkg/operator/k8sutil/service.go
@@ -1,0 +1,63 @@
+/*
+Copyright 2019 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package k8sutil
+
+import (
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// CreateOrUpdateService creates a service or updates the service declaratively if it already exists.
+func CreateOrUpdateService(
+	clientset kubernetes.Interface, namespace string, serviceDefinition *v1.Service,
+) (*v1.Service, error) {
+	name := serviceDefinition.Name
+	logger.Debugf("creating service %s", name)
+	s, err := clientset.CoreV1().Services(namespace).Create(serviceDefinition)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return nil, fmt.Errorf("failed to create service %s. %+v", name, err)
+		}
+		s, err = UpdateService(clientset, namespace, serviceDefinition)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update service %s. %+v", name, err)
+		}
+	}
+	return s, err
+}
+
+// UpdateService updates a service declaratively. If the service does not exist this is considered
+// an error condition.
+func UpdateService(
+	clientset kubernetes.Interface, namespace string, serviceDefinition *v1.Service,
+) (*v1.Service, error) {
+	name := serviceDefinition.Name
+	logger.Debug("updating service %s")
+	existing, err := clientset.CoreV1().Services(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("could not get existing service %s in order to update. %+v", name, err)
+	}
+	// ClusterIP is immutable for k8s services and cannot be left empty in k8s v1 API
+	serviceDefinition.Spec.ClusterIP = existing.Spec.ClusterIP
+	// ResourceVersion required to update services in k8s v1 API to prevent race conditions
+	serviceDefinition.ResourceVersion = existing.ResourceVersion
+	return clientset.CoreV1().Services(namespace).Update(serviceDefinition)
+}

--- a/pkg/operator/k8sutil/statefulset.go
+++ b/pkg/operator/k8sutil/statefulset.go
@@ -47,20 +47,6 @@ func makeHeadlessSvc(name, namespace string, label map[string]string, clientset 
 	return svc, nil
 }
 
-// create a apps.daemonset
-func CreateDaemonSet(name, namespace string, clientset kubernetes.Interface, ds *apps.DaemonSet) error {
-	_, err := clientset.Apps().DaemonSets(namespace).Create(ds)
-	if err != nil {
-		if k8serrors.IsAlreadyExists(err) {
-			_, err = clientset.Apps().DaemonSets(namespace).Update(ds)
-		}
-		if err != nil {
-			return fmt.Errorf("failed to start %s daemonset: %v\n%v", name, err, ds)
-		}
-	}
-	return err
-}
-
 // create a apps.statefulset and a headless svc
 func CreateStatefulSet(name, namespace, appName string, clientset kubernetes.Interface, ss *apps.StatefulSet) (*corev1.Service, error) {
 	label := ss.GetLabels()

--- a/pkg/operator/test/deployment.go
+++ b/pkg/operator/test/deployment.go
@@ -1,0 +1,64 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// WaitForDeploymentImage waits for all deployments with the given labels are running.
+func WaitForDeploymentImage(clientset kubernetes.Interface, namespace, label, container string, initContainer bool, desiredImage string) error {
+
+	sleepTime := 3
+	attempts := 120
+	for i := 0; i < attempts; i++ {
+		deployments, err := clientset.Apps().Deployments(namespace).List(metav1.ListOptions{LabelSelector: label})
+		if err != nil {
+			return fmt.Errorf("failed to list deployments with label %s. %v", label, err)
+		}
+
+		matches := 0
+		for _, d := range deployments.Items {
+			image, err := k8sutil.GetDeploymentSpecImage(clientset, d, container, initContainer)
+			if err != nil {
+				logger.Infof("failed to get image for deployment %s. %+v", d.Name, err)
+				continue
+			}
+			if image == desiredImage {
+				matches++
+			}
+		}
+
+		if matches == len(deployments.Items) && matches > 0 {
+			logger.Infof("all %d %s deployments are on image %s", matches, label, desiredImage)
+			return nil
+		}
+
+		if len(deployments.Items) == 0 {
+			logger.Infof("waiting for at least one deployment to start to see the version")
+		} else {
+			logger.Infof("%d/%d %s deployments match image %s", matches, len(deployments.Items), label, desiredImage)
+		}
+		time.Sleep(time.Duration(sleepTime) * time.Second)
+	}
+	return fmt.Errorf("failed to wait for image %s in label %s", desiredImage, label)
+}

--- a/tests/integration/upgrade_test.go
+++ b/tests/integration/upgrade_test.go
@@ -22,6 +22,7 @@ import (
 
 	opspec "github.com/rook/rook/pkg/operator/ceph/spec"
 	"github.com/rook/rook/pkg/operator/k8sutil"
+	optest "github.com/rook/rook/pkg/operator/test"
 	"github.com/rook/rook/tests/framework/clients"
 	"github.com/rook/rook/tests/framework/installer"
 	"github.com/rook/rook/tests/framework/utils"
@@ -126,7 +127,7 @@ func (s *UpgradeSuite) TestUpgradeToMaster() {
 	require.Nil(s.T(), err)
 
 	// wait for the osd pods to be updated
-	err = k8sutil.WaitForDeploymentImage(s.k8sh.Clientset, s.namespace, "app=rook-ceph-osd", opspec.ConfigInitContainerName, true, "rook/ceph:master")
+	err = optest.WaitForDeploymentImage(s.k8sh.Clientset, s.namespace, "app=rook-ceph-osd", opspec.ConfigInitContainerName, true, "rook/ceph:master")
 	require.Nil(s.T(), err)
 
 	err = s.k8sh.WaitForLabeledPodsToRun("app=rook-ceph-osd", s.namespace)

--- a/tests/integration/upgrade_test.go
+++ b/tests/integration/upgrade_test.go
@@ -119,9 +119,8 @@ func (s *UpgradeSuite) TestUpgradeToMaster() {
 	assert.Nil(s.T(), err)
 	assert.Equal(s.T(), "rook/ceph:"+installer.VersionMaster, version)
 
-	// wait for the mon pods to be running
-	err = k8sutil.WaitForDeploymentImage(s.k8sh.Clientset, s.namespace, "app=rook-ceph-mon", opspec.ConfigInitContainerName, true, "rook/ceph:master")
-	require.Nil(s.T(), err)
+	// wait for the mon pods to be running. we can no longer check a version in the pod spec, so we just wait a bit.
+	time.Sleep(15 * time.Second)
 
 	err = s.k8sh.WaitForLabeledPodsToRun("app=rook-ceph-mon", s.namespace)
 	require.Nil(s.T(), err)
@@ -160,7 +159,7 @@ func (s *UpgradeSuite) updateClusterRoles() error {
 	if err := s.k8sh.DeleteResource("ClusterRole", "rook-ceph-cluster-mgmt"); err != nil {
 		return err
 	}
-	if err := s.k8sh.DeleteResource("-n", s.namespace, "Role", "rook-ceph-system"); err != nil {
+	if err := s.k8sh.DeleteResource("-n", systemNamespace, "Role", "rook-ceph-system"); err != nil {
 		return err
 	}
 	if err := s.k8sh.DeleteResource("-n", s.namespace, "Role", "rook-ceph-mgr-system"); err != nil {
@@ -281,7 +280,7 @@ apiVersion: rbac.authorization.k8s.io/v1beta1
 kind: Role
 metadata:
   name: rook-ceph-system
-  namespace: ` + s.namespace + `
+  namespace: ` + systemNamespace + `
   labels:
     operator: rook
     storage-backend: ceph
@@ -304,17 +303,6 @@ rules:
   - apps
   resources:
   - daemonsets
-  verbs:
-  - get
-  - list
-  - watch
-  - create
-  - update
-  - delete
-- apiGroups:
-  - apps
-  resources:
-  - daemonsets
   - statefulsets
   verbs:
   - get
@@ -328,7 +316,6 @@ kind: ClusterRole
 apiVersion: rbac.authorization.k8s.io/v1beta1
 metadata:
   name: rook-ceph-mgr-system
-  namespace: ` + s.namespace + `
 rules:
 - apiGroups:
   - ""

